lib/Dialect/AIEVec/IR/AIEVecOps.cpp

```diff
@@ -1284,5 +1284,272 @@ ParseResult UnpackOp::parse(OpAsmParser &parser, OperationState &result) {
+template <>
+inline void printAccumulator(OpAsmPrinter &p, aievec::FMAConvOp op) {
+  p << ", " << op.getAcc();
+}
+template <>
+inline void printAccumulator(OpAsmPrinter &p, aievec::MulConvOp op) {}
+
+// Mark fmsub indicator as elided if the FMAElem op is not fmsub
+template <typename T>
+inline void elideFMSubAttr(T op, SmallVector<StringRef, 4> &elidedAttrs);
+template <>
+inline void elideFMSubAttr(aievec::FMAConvOp op,
+                           SmallVector<StringRef, 4> &elidedAttrs) {
+  if (!op.getFmsub())
+    elidedAttrs.push_back(op.getSubAttrName());
+}
+
+template <>
+inline void elideFMSubAttr(aievec::MulConvOp op,
+                           SmallVector<StringRef, 4> &elidedAttrs) {}
+
+// Print out MulConv and FMAConv op.
+template <typename T> static void printMulFMAConvOp(OpAsmPrinter &p, T op) {
+  // Print the left operand
+  p << " " << op.getLhs();
+  // Print the right operand
+  p << ", " << op.getRhs();
+  // For fma op, print the accumulator
+  printAccumulator(p, op);
+
+  // Print the attributes, but don't print attributes that are empty strings
+  SmallVector<StringRef, 4> elidedAttrs;
+  for (int idx = 0; idx < 2; ++idx) {
+    elideFMSubAttr(op, elidedAttrs);
+  }
+  p.printOptionalAttrDict(op->getAttrs(), elidedAttrs);
+
+  // And now print the types
+  p << " : " << op.getLhs().getType() << ", " << op.getRhs().getType();
+  p << ", " << op.getResult().getType();
+}
+
+void MulConvOp::print(OpAsmPrinter &p) {
+  printMulFMAConvOp<aievec::MulConvOp>(p, *this);
+}
+
+void aievec::FMAConvOp::print(OpAsmPrinter &p) {
+  printMulFMAConvOp<aievec::FMAConvOp>(p, *this);
+}
+
+// Verify MulConv and FMAConv op.
+template <typename T> LogicalResult verifyMulFMAConvOp(T op) {
+  // Verify the types
+  VectorType lhsType = op.getLhs().getType().template dyn_cast<VectorType>();
+  VectorType rhsType = op.getRhs().getType().template dyn_cast<VectorType>();
+
+  if (!lhsType || !rhsType)
+    return op.emitError("requires vector type");
+
+  VectorType resultType =
+      op.getResult().getType().template dyn_cast<VectorType>();
+
+  if (!resultType)
+    return op.emitError("requires vector type");
+
+  // Additional checks for FMAElem op
+  // Get the width of the underlying scalars of all the vectors
+  Type ltype = lhsType.getElementType();
+  Type rtype = rhsType.getElementType();
+  Type atype = resultType.getElementType();
+
+  if (!ltype.isa<IntegerType>() || !rtype.isa<IntegerType>() ||
+      !atype.isa<IntegerType>()) {
+    return op.emitError("requires integer type");
+  }
+
+  unsigned ltypeWidth = ltype.getIntOrFloatBitWidth();
+  unsigned rtypeWidth = rtype.getIntOrFloatBitWidth();
+  unsigned atypeWidth = atype.getIntOrFloatBitWidth();
+
+  // Checks on the number of lanes
+  unsigned accLanes = getVectorLaneSize(resultType);
+  unsigned rhsLanes = getVectorLaneSize(rhsType);
+  unsigned lhsLanes = getVectorLaneSize(lhsType);
+
+  // lane size must match
```

Comment:
There are some basic requirements on M,N:
M > 0, 
N > 0
2 * M >= M + N - 1 (because we will be shifting lhs over N-1 positions

Furthermore: 
accLanes == M
lhsLanes == M * 2

Reply:
This is a good point!